itemloaders/__init__.py

```diff
@@ -64,6 +64,11 @@ class ItemLoader:
         An Item class (or factory), used to instantiate items when not given in
         the ``__init__`` method.
 
+        .. warning:: This factory/class needs to be callable/instantiated without
+            any arguments.
+            When working with ``dataclass`` you may need to decorate your
+            classes using ``@dataclass(init=False)``.
```

Comment:
Hm, that's an interesting workaround. It means that user can't create `Product(name='Plasma Tv', price=999.98)` manually. I'm not sure I understand all consequences. It seems we don't test `init=False` dataclasses in itemadapter (//cc @elacuesta), I'm not sure what's the behavior e.g. with serialization, if fields are missing.

Overall, this looks like an issue in itemloader which can be fixed, not a problem with dataclasses or itemadapter. Fixing it is out of scope for this PR though.

What do you think about asking to provide default field values instead, and saying "currently" when explaining this limitation? We may want to lift it in future.

Reply:
Agree with _Currently_.
However, I'm not sure about the default field values..
Do you have something in mind? Like an extra argument or similar?

This problem only happens because we try to create an empty item to set in the context and to return `item` property.
If we didn't need to set in the context, we could avoid creating it in the initialization and create it only if required.

I created some tests here
```
from dataclasses import dataclass
from itemadapter import ItemAdapter

@dataclass
class Product:
    name: str
    price: float

p = Product()  # error, needs the fields
p = Product(name=None, price=None)
values = ItemAdapter(p).items()
list(values)
#  [('name', None), ('price', None)]

# EXAMPLE 2

from dataclasses import dataclass
from itemadapter import ItemAdapter

@dataclass(init=False)
class Product:
    name: str
    price: float

p = Product()
values = ItemAdapter(p).items()  # error
list(values)
# []
print(values)  # raises an error

```

Reply:
I've created an issue for this here https://github.com/scrapy/itemloaders/issues/14

Reply:
I think `dataclass` methods expect all fields to be set, either by the module-provided `__init__` method, by a custom one if the user sets `init=False`, or one by one at a later stage. For instance, the following example fails until all fields are manually set:

```python
>>> from dataclasses import dataclass, asdict
>>> @dataclass(init=False)
... class Product:
...     name: str
...     price: float
... 
>>> p = Product()
>>> p
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dataclasses.py", line 357, in wrapper
    result = user_function(self)
  File "<string>", line 2, in __repr__
AttributeError: 'Product' object has no attribute 'name'
>>> asdict(p)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dataclasses.py", line 1044, in asdict
    return _asdict_inner(obj, dict_factory)
  File "/usr/local/lib/python3.7/dataclasses.py", line 1051, in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
AttributeError: 'Product' object has no attribute 'name'
>>> p.name = "foo"
>>> p
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dataclasses.py", line 357, in wrapper
    result = user_function(self)
  File "<string>", line 2, in __repr__
AttributeError: 'Product' object has no attribute 'price'
>>> asdict(p)     
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dataclasses.py", line 1044, in asdict
    return _asdict_inner(obj, dict_factory)
  File "/usr/local/lib/python3.7/dataclasses.py", line 1051, in _asdict_inner
    value = _asdict_inner(getattr(obj, f.name), dict_factory)
AttributeError: 'Product' object has no attribute 'price'
>>> p.price = 5
>>> p
Product(name='foo', price=5)
>>> asdict(p)  
{'name': 'foo', 'price': 5}
```

I think using `init=False` and setting all fields in the loader is probably more error-prone that defining default field values.

Reply:
Based on @elacuesta comment and on his PR https://github.com/scrapy/scrapy/pull/3881
Here are a few examples I could put together..

## Default

```
@dataclass
class Product1:
    name: str
    price: float


class Product1Loader(ItemLoader):
    default_item_class = Product1
    default_output_processor = TakeFirst()
    name_out = Join(' -- ')

p1l = Product1Loader(
    item=Product1(name='product 1', price=9)
)
p1l.add_value('name', 'new product')
print(p1l.load_item())
# Product1(name='product 1 -- new product', price=9)

p1l = Product1Loader()
# Traceback (most recent call last):
#   File "examples.py", line 17, in <module>
#     p1l = Product1Loader()
#   File "/home/julio/Projects/scrapy-itemloaders/itemloaders/__init__.py", line 99, in __init__
#     item = self.default_item_class()
# TypeError: __init__() missing 2 required positional arguments: 'name' and 'price'
```

## init=False
```
@dataclass(init=False)
class Product2:
    name: str
    price: float

class Product2Loader(ItemLoader):
    default_item_class = Product2
    default_output_processor = TakeFirst()
    name_out = Join(' -- ')

p = Product2()
p.name = 'product 1'
p.price = 9
p2l = Product2Loader(item=p)
p2l.add_value('name', 'new product')
print(p2l.load_item())
# Product2(name='product 1 -- new product', price=9)

p2l = Product2Loader()
p2l.add_value('name', 'new product')
print(p2l.load_item())  # "loaded properly", but __repr__ requires the missing field
# Traceback (most recent call last):
#   File "examples.py", line 19, in <module>
#     print(p2l.load_item())
#   File "/usr/lib/python3.7/dataclasses.py", line 357, in wrapper
#     result = user_function(self)
#   File "<string>", line 2, in __repr__
# AttributeError: 'Product2' object has no attribute 'price'
```

## field default
```
@dataclass
class Product3:
    name: str = field(default=None)
    price: float = field(default=None)

class Product3Loader(ItemLoader):
    default_item_class = Product3
    default_output_processor = TakeFirst()
    name_out = Join(' -- ')

p3l = Product3Loader(item=Product3(name='product 1', price=9))
p3l.add_value('name', 'new product')
print(p3l.load_item())
# Product3(name='product 1 -- new product', price=9)

p3l = Product3Loader()
p3l.add_value('name', 'new product')
print(p3l.load_item())
# Product3(name='new product', price=None)
```

## field factory
```
@dataclass
class Product4:
    name: str = field(default_factory=list)
    price: float = field(default_factory=list)

class Product4Loader(ItemLoader):
    default_item_class = Product4
    default_output_processor = TakeFirst()
    name_out = Join(' -- ')

p4l = Product4Loader(item=Product4(name='product 1', price=9))
p4l.add_value('name', 'new product')
print(p4l.load_item())
# Product4(name='product 1 -- new product', price=9)

p4l = Product4Loader()
p4l.add_value('name', 'new product')
print(p4l.load_item())
# Product4(name='new product', price=[])  # here price is a list
```

## field init=False (similar to class init=False)
```
@dataclass
class Product5:
    name: str = field(init=False)
    price: float = field(init=False)

class Product5Loader(ItemLoader):
    default_item_class = Product5
    default_output_processor = TakeFirst()
    name_out = Join(' -- ')

p = Product5()
p.name = 'product 1'
p.price = 9
p5l = Product5Loader(item=p)
p5l.add_value('name', 'new product')
print(p5l.load_item())
# Product5(name='product 1 -- new product', price=9)

p5l = Product5Loader()
p5l.add_value('name', 'new product')
print(p5l.load_item())  # "loads properly", but __repr__ requires the field
# Traceback (most recent call last):
#   File "examples.py", line 19, in <module>
#     print(p5l.load_item())
#   File "/usr/lib/python3.7/dataclasses.py", line 357, in wrapper
#     result = user_function(self)
#   File "<string>", line 2, in __repr__
# AttributeError: 'Product5' object has no attribute 'price'
```

It seems to me that `default` or `default_factory` are the way to go.
Even in both cases, we can agree that, we could identify it as a dataclass in `ItemLoader` and init it with all fields set to `None` or `[]` to avoid having this burden in the user.

Reply:
> we could identify it as a dataclass in `ItemLoader` and init it with all fields set to `None` or `[]`

Nice, that would be a good solution I think. You could use `itemadapter.utils.is_dataclass_instance` and `itemadapter.ItemAdapter::field_names` to help with that.

Reply:
Alternative is to work on dicts internally, and create an item only when .load_item is called. This would ensure that an error is raised for non-optional fields, as it should. Though this is a much larger change, which need to be made carefully.

Reply:
@ejulio the examples in "field factory" and "field default" should be a bit more complex, as types are not correct. It should be Optional[...] in "field default", and probably `List[...]`  in field factory.

Reply:
Indeed, even though it works :thinking: 
About the change to use `dict` internally, we expose it to the outside as a property `item` in `ItemLoader`.
So, need to be careful that it will change the interface a bit..
Even though, so far, the interfaces rely mainly in scrapy Item and it is similar to a dict, so it shouldn't make that much of a difference.
However, it could awkward to use a dict syntax if your model is a dataclass

Reply:
> About the change to use dict internally, we expose it to the outside as a property item in ItemLoader.
So, need to be careful that it will change the interface a bit..

Not necessarily - .item property can be computed on fly. What I'm proposing is that .load_item() returns an item of the type you asked, but if a dataclass or attr.s is `default_item_class`, then their field meta is inspected, but the item is actually instantiated when .load_item() is called, not immediately. 

From the user point of view the API should stay the same (I hope - haven't looked in detail), but the item of default_item_class is only created towards the end of the extraction, not in the beginning of the extraction. This allows item to exist in "partial", unfinished stage, until extraction is done and .load_item is called. 

In this case it can be good that creation of the item fails on missing fields, e.g. because some required field was not added via .add_value or similar methods.

This change requires careful inspection of the Item Loaders API, tests, documentation, etc. - it is a huge change, which we would have to try hard to make backwards compatible. I think it is out of scope for this PR.

---

So, let's say after a bit of discussion we agree that such change is desirable, and we're going to support dataclasses with required fields in future, by deferring their creation. What to do now in this case?

My proposal is to leave this problem aside, as much as possible. Document a simple workaround (default fields), say that currently itemloaders create instances of `default_item_class` early in the process, so they need to be "creatable" without arguments, but that users shouldn't be relying on that, as it may change in future. Maybe declare dataclasses and attr.s support experimental.

I'm not sure init=False workaround is good. The intended use of this argument is to create classes with custom `__init__` methods, not to allow half-broken dataclass instances. `init=False` fields also need to be used with care (a quote from https://www.python.org/dev/peps/pep-0557/):

>  It is expected that init=False fields will be rarely and judiciously used. 

For example, if I read PEP-557 correctly, `.replace()` is broken for `init=False` fields. For classes with `init=False` we've seen `__repr__` broken. 

Using fields with default values make it explicit that an item can be in a partial stage, and that additional validation may be required if you intend some fields to be non-optional. Because of that using default values sounds better to me, even if it is a bit more boilerplate.

Reply:
Agreed. I'll remove the `init=False` from the docs then.